### PR TITLE
Garbage Collection in `execute_computation_action`, Fix Inplace-Operation Identification

### LINKED ISSUE #3397: Memory leak in F.relu() method in AST
This code, causes a memory leak only on the first worker in the worker list passed to the `encrypt()`. In this example it is  'alice', the second worker 'bob' does not suffer from any leaks.

The reason is the `F.relu()` method:

```
import torch as th
import torch.optim as optim
import pprint
import syft as sy
import torch.nn.functional as F

hook = sy.TorchHook(th) #, verbose=True)
bob = sy.VirtualWorker(hook, id="bob") #, verbose=True)
alice = sy.VirtualWorker(hook, id="alice") #, verbose=True)
crypto_provider = sy.VirtualWorker(hook, id="james") #, verbose=True)



a = th.ones(1, 5)
a = a.encrypt(workers=[alice, bob], crypto_provider=crypto_provider, requires_grad=True)


class Classifier(th.nn.Module):
    def __init__(self, in_features, out_features):
        super(Classifier, self).__init__()
        self.fc = th.nn.Linear(in_features, out_features)

    def forward(self, x):
        logits = self.fc(x)
        logits = F.relu(logits)
        return logits


classifier = Classifier(in_features=5, out_features=5)
model = classifier.fix_prec().share(alice, bob, crypto_provider=crypto_provider, requires_grad=True)

for i in range(3):
    print("=" * 10 + f"Iter{i + 1}" + "=" * 10)
    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")

    preds = classifier(a)


    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")

print("~" * 10 + "Done" + "~" * 10)
```

Here is the output (Only Alice leaks)

```
==========Iter1==========
Alice: 3
Bob: 3
Alice: 11
Bob: 7
==========Iter2==========
Alice: 11
Bob: 7
Alice: 15
Bob: 7
==========Iter3==========
Alice: 15
Bob: 7
Alice: 19
Bob: 7
~~~~~~~~~~Done~~~~~~~~~~

```

If you comment out `logits = F.relu(logits)`, no memory leaks are observed:

```
==========Iter1==========
Alice: 3
Bob: 3
Alice: 3
Bob: 3
==========Iter2==========
Alice: 3
Bob: 3
Alice: 3
Bob: 3
==========Iter3==========
Alice: 3
Bob: 3
Alice: 3
Bob: 3
~~~~~~~~~~Done~~~~~~~~~~

```

### LINKED ISSUE #3398: Memory leak in the argmax() method of AST
This code, causes a memory leak only on the first worker in the worker list passed to the `encrypt()`. In this example it is  'alice', the second worker 'bob' does not suffer from any leaks.

The reason is the `argmax()` method:

```
import torch as th
import torch.optim as optim
import syft as sy

hook = sy.TorchHook(th) #, verbose=True)
bob = sy.VirtualWorker(hook, id="bob") #, verbose=True)
alice = sy.VirtualWorker(hook, id="alice") #, verbose=True)
crypto_provider = sy.VirtualWorker(hook, id="james") #, verbose=True)

a = th.ones(1, 5)
a = a.encrypt(workers=[alice, bob], crypto_provider=crypto_provider, requires_grad=True)

for i in range(3):
    print("=" * 10 + f"Iter{i + 1}" + "=" * 10)
    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")

    a.argmax(dim=1)

    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")

print("~" * 10 + "Done" + "~" * 10)
```

Here is the output (Only Alice leaks):

```
==========Iter1==========
Alice: 1
Bob: 1
Alice: 13
Bob: 1
==========Iter2==========
Alice: 13
Bob: 1
Alice: 25
Bob: 1
==========Iter3==========
Alice: 25
Bob: 1
Alice: 37
Bob: 1
~~~~~~~~~~Done~~~~~~~~~~
```

If you comment out `a.argmax(dim=1)`, no memory leaks are observed:

```
==========Iter1==========
Alice: 1
Bob: 1
Alice: 1
Bob: 1
==========Iter2==========
Alice: 1
Bob: 1
Alice: 1
Bob: 1
==========Iter3==========
Alice: 1
Bob: 1
Alice: 1
Bob: 1
~~~~~~~~~~Done~~~~~~~~~~



```

### PR DESCRIPTION
**Pull request by Syzygianinfern0**
## Description

Fixes #3397 
Fixes #3398 

~Currently breaks inplace computations. WIP 🚧🚧~
- [x] GC
- [x] Fix inplace operations

## Type of change

Please mark options that are relevant.

- [ ] Added/Modified tutorials
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

* [x] I have added tests for my changes
* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

### CODE REVIEW

**Comment by AlanAboudib** [comment]
@LaRiffle would you please review this quickly :)

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 1]
```diff
@@ -943,7 +943,7 @@ def max(self, dim=None, return_idx=False):
             the maximum value (possibly across an axis)
             and optionally the index of the maximum value (possibly across an axis)
         """
-        values = self
+        values = self.copy()
```
Comment (karlhigley) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:946]: Does this inherently require copying because operations later in this function are in-place? Could it be avoided by changing the lines below that read something like `values = values.some_op()`?

File: syft/frameworks/torch/torch_attributes.py [hunk 2]
```diff
@@ -135,6 +133,12 @@ def __init__(self, torch: ModuleType, hook: ModuleType) -> None:
 
         # Dict {method_name: <is_inplace:bool>
         self.inplace_methods = {}
+        self._inplace_pattern = re.compile(r"(^__i(?!nit|mport|ter).+_)|^((?!^_+).+)_$")
```
Comment (karlhigley) [inline syft/frameworks/torch/torch_attributes.py:136]: Could you add a test for this regex?

File: syft/generic/frameworks/hook/hook_args.py [hunk 3]
```diff
@@ -717,6 +717,9 @@ def register_tensor(
         response_ids: List of ids where the tensor should be stored
             and each id is pop out when needed.
     """
+    # `de_register` initially as this causes the original PT
```
Comment (karlhigley) [inline syft/generic/frameworks/hook/hook_args.py:720]: What causes the original PT not to get GC'd? Unclear from this comment.

File: syft/workers/base.py [hunk 4]
```diff
@@ -492,6 +492,7 @@ def execute_computation_action(self, action: ComputationAction) -> PointerTensor
             if type(_self) == str and _self == "self":
                 _self = self
             if sy.framework.is_inplace_method(op_name):
+                # TODO [sharan]: Can this be used to fix the inplace operations breakages??
```
Comment (karlhigley) [inline syft/workers/base.py:495]: If there's work left to be done, better to create an issue than a TODO.

Comment (Syzygianinfern0) [reply syft/workers/base.py]: Removed (it was addressed)

File: syft/generic/frameworks/hook/hook_args.py (hunk 3 above)
Comment (Syzygianinfern0) [reply syft/generic/frameworks/hook/hook_args.py]: Updated :smile:

File: syft/frameworks/torch/torch_attributes.py (hunk 2 above)
Comment (Syzygianinfern0) [reply syft/frameworks/torch/torch_attributes.py]: Done :+1:

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 1 above)
Comment (Syzygianinfern0) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py]: Removed (redundant operation)

File: syft/generic/frameworks/hook/hook_args.py [hunk 5]
```diff
@@ -717,6 +717,11 @@ def register_tensor(
         response_ids: List of ids where the tensor should be stored
             and each id is pop out when needed.
     """
+    # This method often leads to re-registration of tensors
+    # hence creating two copies of the same info. The older tensor
+    # is left hanging and is never deleted. De-Registering the original
+    # tensor (if-exists) before registration addresses this problem.
+    owner.de_register_obj(tensor)  # Doesn't raise Exceptions if absent on owner
```
Comment (karlhigley) [inline syft/generic/frameworks/hook/hook_args.py:724]: 👍
